File: JSON/CRUD_simples/crud.js
const fs = require("fs");
const prompt = require("prompt-sync")();

// Função para criar o menu de opções:
function menu() {
    console.log("\nMenu de Contatos: ");
    console.log("1. Adicionar Contato.")
    console.log("2. Listar Contatos.");
    console.log("3. Atualizar Contato.");
    console.log("4. Excluir Contato.")
    console.log("5. Sair.")
}

//Função principal para executar o programa
function main() {
    let opcao;
    do {
        menu();
        opcao = prompt("Escolha uma opção: ");
        switch (opcao) {
            case "1":
                adicionar(); //Chama a função para adicionar um novo contato
                break;
            case "2":
                listar(); //Chama a função para listar os contatos
                break;
            case "3":
                atualizar(); //Chama a que atualiza a lista de contatos
                break;
            case "4":
                excluir(); //Chama a função que apaga um contato da lista
                break;
            case "5":
                console.log("Saindo do programa...");
                break;
            default:
                console.log("Opção inválida. Tente novamente.");
        }
    } while (opcao !== "5");
}

// Função para ler os dados do arquivo JSON
function lerDados() {
    const dados = fs.readFileSync("contatos.json", "utf-8");
    return JSON.parse(dados || "[]"); // Retorna um array vazio se não houver dados no JSON
}

//Função para adicionar um novo contato
function adicionar() {
    const nome = prompt("Digite o nome do contato: ");
    const telefone = prompt("Digite o telefone do contato: ");

    const contatos = lerDados();
    contatos.push({nome, telefone});
    salvarDados(contatos);
    console.log("\nContato adicionado.");
}

//Função para listar os contatos
function listar() {
    const contatos = lerDados();
    console.log("Lista de Contatos: ");
    contatos.forEach((contato, index) => {
        console.log(`${index+1}. ${contato.nome} = ${contato.telefone}`);
    })
}

//Função para atualizar um contato
function atualizar() {
    const indexAtualizar = parseInt(prompt("Digite o ID do contato a ser atualizado: ")) -1;
    const contatos = lerDados();
    if (indexAtualizar >= 0 && indexAtualizar < contatos.length) {
        const novoNome = prompt("Digite o novo nome do contato: ");
        const novoTelefone = prompt("Digite o novo número de telefone: ");

        contatos[indexAtualizar] = {nome: novoNome, telefone: novoTelefone};
        salvarDados(contatos);
        console.log("\nContato atualizado.");
    } else {
        console.log("\nID do contato inválido.");
    }
}

//Função para excluir um contato
function excluir() {
    const indexExcluir = parseInt(prompt("Digite o ID do contato a ser excluido: ")) -1;
    const contatos = lerDados();
    if (indexExcluir >= 0 && indexExcluir < contatos.length) {
        contatos.splice(indexExcluir, 1);
        salvarDados(contatos);
        console.log("\nContato excluído.");
    } else {
        console.log("ID do contato inválido.");
    }
}

// Função para salvar/gravar os dados no arquivo JSON
function salvarDados(contatos) {
    fs.writeFileSync("contatos.json", JSON.stringify(contatos, null, 2));
}

//Inicia o programa 
listar()
main();

File: JSON/crud_json/crud.py
import json 

# --- FUNÇÕES DE INTERFACE ---

# Exibe o menu principal de opções para o usuário
def menu():
    print("\n=== TCG - Treinadores e Pokémons ===")
    print("1. Adicionar.")
    print("2. Listar.")
    print("3. Atualizar.")
    print("4. Excluir.")
    print("5. Sair.")

# Gerencia a escolha de categoria (granularidade) para as operações
def escolher_grupo():
    #pergunta ao usuário o grupo que deseja retornar
    print("\nTreinador ou Pokémon: ")
    print("1. Treinador")
    print("2. Pokémon")

    opcao = input("Escolha: ")

    # Retorna a chave correspondente ao dicionário no JSON
    if opcao == "1":
        return "treinadores"
    
    elif opcao == "2":
        return "pokemons"
    
    else:
        print("Opção Inválida.")
        return None
    
# --- FUNÇÕES DE PERSISTÊNCIA DE DADOS (JSON) ---

# Carrega os dados do arquivo físico para a memória do programa
def ler_dados():
    # Abre o arquivo em modo de leitura ('r') com suporte a acentos (utf-8)
    with open("pokemon.json", "r", encoding="utf-8") as arquivo:
        return json.load(arquivo)
    
# Grava os dados da memória de volta para o arquivo físico
def salvar_dados(dados):
    # Abre o arquivo em modo de escrita ('w') formatando o JSON para leitura fácil
    with open("pokemon.json", "w", encoding="utf-8") as arquivo:
        json.dump(dados, arquivo, indent=2, ensure_ascii=False)

# --- OPERAÇÕES DO CRUD ---

# CREATE: Adiciona um novo registro ao grupo selecionado
def adicionar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    nome = input("Nome: ")
    tipo = input("Tipo: ")

    dados = ler_dados() # Lê os dados atuais antes de modificar

    # Adiciona o novo dicionário à lista do grupo escolhido
    dados[grupo].append({
        "nome": nome,
        "tipo": tipo
    })

    # Persiste a alteração no arquivo
    salvar_dados(dados)
    print("Personagem adicionado.")

# READ: Exibe os registros de um grupo específico
def listar():
    grupo = escolher_grupo()
    
    if not grupo:
        return
    
    dados = ler_dados()
    print(f"\nLista de {grupo.upper()}: ")

    # Percorre a lista do grupo usando enumerate para gerar IDs visuais (index)
    for index, pokemon in enumerate(dados[grupo], start=1):
        print(f"{index}. {pokemon['nome']} - {pokemon['tipo']}")

# UPDATE: Altera os dados de um registro existente via índice
def atualizar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()

    # Pede o índice ao usuário e ajusta para o padrão de lista (começando em 0)
    index = int(input("Indíce do personagem: ")) - 1

    # Validação de segurança para garantir que o índice existe na lista
    if 0 <= index < len(dados[grupo]):
        nome = input("Novo nome: ")
        tipo = input("Novo tipo: ")

        # Substitui os dados antigos pelos novos
        dados[grupo][index] = {
            "nome": nome,
            "tipo": tipo
        }
    
        salvar_dados(dados)
        print("Personagem atualizado.")

    else:
        print("Indíce inválido.")

# DELETE: Remove um registro do arquivo JSON
def excluir():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Indíce do personagem: ")) - 1

    # Verificar se o index é válido antes de tentar remover
    if 0 <= index < len(dados[grupo]):
        # Remove o elemento da lista pelo índice usando pop()
        dados[grupo].pop(index)

        salvar_dados(dados)
        print("Personagem excluído.")
    else:
        print("Indíce inválido.")

# --- FLUXO PRINCIPAL ---

# Função que mantém o programa em execução e gerencia as chamadas do menu
def main():
    while True:
        menu()

        opcao = input("\nEscolha uma opção: ")

        # Estrutura de decisão para chamar as funções do CRUD
        if opcao == "1":
            adicionar()

        elif opcao == "2":
            listar()

        elif opcao == "3":
            atualizar()

        elif opcao == "4":
            excluir()

        elif opcao == "5":
            print("Encerrando o programa...")
            break

        else:
            print("Opção inválida.")

# Executa o programa
main()